src/http-api/interface.ts

```diff
@@ -18,6 +18,13 @@ import { MatrixError } from "./errors";
 
 export type Body = Record<string, any> | BodyInit;
 
+/**
+ * @experimental
+ */
+export type TokenRefreshFunction = (refreshToken: string) => Promise<{
+    accessToken: string;
+    refreshToken?: string;
+}>;
```

Comment:
I think it would be nicer to express this as an interface (which `OidcTokenRefresher` would implement):

```suggestion
export interface TokenRefresher {
    doRefreshAccessToken(refreshToken: string): Promise<{
        accessToken: string;
        refreshToken?: string;
    }>;
}
```

You could then just pass the `tokenRefresher` instance into `ICreateClientOpts` rather than having to mess about with `bind` [here](https://github.com/matrix-org/matrix-react-sdk/pull/11699/files#diff-d3f0f3af9b2a4141351677842e0b40f673397ca63c3292c8bf98294149ff11a4R388).

Reply:
Also, I know it's `@experimental` but I still think it's worth documenting.

Reply:
Also also: I'd give the return type its own `interface` definition so you don't need to refer to it as `ReturnType<...>`, but YMMV on that

Reply:
I think there is some clarity in using `ReturnType<TokenRefreshFunction>` because it makes it clear it is implementing that function type

Reply:
Disagree about the interface, it needlessly complicates the refresh api within `http-api`.

Reply:
> I think there is some clarity in using `ReturnType<TokenRefreshFunction>` because it makes it clear it is implementing that function type

well no, it says it *returns* the same thing as a `TokenRefreshFunction`, but says nothing about the parameters.

TBH if your goal is to make it clear that a function is implementing a `TokenRefreshFunction`, then IMHO the clearest way to do that is to write it in a (doc-)comment.

Thanks for pulling out the `AccessTokens` interface.

Reply:
> Disagree about the interface, it needlessly complicates the refresh api within `http-api`.

I don't really see what you mean, but fine, I don't feel that strongly: let's leave it as a function type if you prefer.